### DISCUSSION TITLE
Write string to file

### DISCUSSION THREAD

**Discussion by gregnordin** [pyscript/ltk#17] [General]
Is there a way to write a string to a file with LTK? I can do it with pyscript (example below), but if it can be done with LTK, it would make it more convenient for the small app I plan to develop. I have written a finite element simulation tool that I am adapting to use  TOML files to specify and control the simulation geometry and specifics. The small app I want to develop with LTK provides a user interface to set up the geometry and simulation details. As such it will need to read and write TOML files, which is why I ask about using LTK to write a string to a file, which is the last base capability that I need. If there isn't a way to do this with LTK, I'll plan on having part of the app use pyscript directly for this feature. Ultimately I would like to open source the whole simulation tool and TOML file setup UI app to benefit my corner of the scientific world.

`index.html` file with pyscript code to write to a string to a file:
```
# index.html
<!DOCTYPE html>
<html>
<head>
    <title>PyScript String Writer</title>
    <script defer src="https://pyscript.net/latest/pyscript.js"></script>
    <link rel="stylesheet" href="https://pyscript.net/latest/pyscript.css">
</head>
<body>
    <button id="saveButton">Save Example String</button>

    <py-config>
        packages = []
    </py-config>

    <py-script>
from js import document, window, Blob, URL, console
from pyodide.ffi import create_proxy

def write_string_to_file(content, filename):
    try:
        # Create a Blob containing the text
        blob = Blob.new([content], {type: 'text/plain'})
        
        # Create a URL for the Blob
        url = URL.createObjectURL(blob)
        
        # Create a temporary link element
        link = document.createElement('a')
        link.href = url
        link.download = filename
        
        # Append link to body, click it, and remove it
        document.body.appendChild(link)
        link.click()
        document.body.removeChild(link)
        
        # Clean up the URL
        URL.revokeObjectURL(url)
        
    except Exception as e:
        console.error(f"Error saving file: {str(e)}")

def save_example(event):
    # Example string
    my_string = """This is an example string
with multiple lines
of text content."""
    
    write_string_to_file(my_string, "example.txt")

# Add event listener to the save button
save_button = document.getElementById('saveButton')
save_button.addEventListener('click', create_proxy(save_example))
    </py-script>
</body>
</html>
```

**Comment by laffra** [comment]
I added a class method to `ltk.File` that makes downloading a file easier than the code above. Click the "load a file" link at [LTK Kitchensink](http://127.0.0.1:3002/index.html?tab=2). It will upload a file to the browser and then download it directly using `ltk.File.download`. The file type is derived from the file itself, so it works for text and binary files, but you can also override the content type.

Also, check out the [Python context manager](http://127.0.0.1:3002/index.html?tab=2) used by `ltk.File` for creating the object URL and ensure the URL is always revoked.

**Comment by gregnordin** [comment]
Awesome, again! I've played around with `ltk.File.download` and gone through the relevant code. The context manager for [`object_url`](https://github.com/pyscript/ltk/blob/main/ltk/jquery.py#L142) looks like it does the right thing and is working correctly in my tests. I created an [example on pyscript.com](https://pyscript.com/@gnordin/ltk-write-text-to-file/latest) in case it is helpful.
